File: mobile/library/swift/Stream.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// A type representing a stream that is actively transferring data.
///
/// Constructed using `StreamPrototype`, and used to write to the network.
@objcMembers
public class Stream: NSObject {
  private let underlyingStream: EnvoyHTTPStream

  /// Initialize a new instance of the stream.
  ///
  /// - parameter underlyingStream: Underlying stream that can be used to send request data.
  init(underlyingStream: EnvoyHTTPStream) {
    self.underlyingStream = underlyingStream
    super.init()
  }

  // MARK: - Public

  /// Send headers over the stream.
  ///
  /// - parameter headers:   Headers to send over the stream.
  /// - parameter endStream: Whether this is a headers-only request.
  ///
  /// - returns: This stream, for chaining syntax.
  @discardableResult
  public func sendHeaders(_ headers: RequestHeaders, endStream: Bool) -> Stream {
    self.underlyingStream.sendHeaders(headers.headers, close: endStream)
    return self
  }

  /// Send data over the stream.
  ///
  /// - parameter data: Data to send over the stream.
  ///
  /// - returns: This stream, for chaining syntax.
  @discardableResult
  public func sendData(_ data: Data) -> Stream {
    self.underlyingStream.send(data, close: false)
    return self
  }

  /// Close the stream with trailers.
  ///
  /// - parameter trailers: Trailers with which to close the stream.
  public func close(trailers: RequestTrailers) {
    self.underlyingStream.sendTrailers(trailers.headers)
  }

  /// Close the stream with a data frame.
  ///
  /// - parameter data: Data with which to close the stream.
  public func close(data: Data) {
    self.underlyingStream.send(data, close: true)
  }

  /// Cancel the stream.
  public func cancel() {
    _ = self.underlyingStream.cancel()
  }
}

File: mobile/library/kotlin/io/envoyproxy/envoymobile/Stream.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyHTTPStream
import java.nio.ByteBuffer

/**
 * A type representing a stream that is actively transferring data.
 *
 * Constructed using `StreamPrototype`, and used to write to the network.
 */
open class Stream(
  private val underlyingStream: EnvoyHTTPStream,
  private val useByteBufferPosition: Boolean
) {
  /**
   * Send headers over the stream.
   *
   * @param headers Headers to send over the stream.
   * @param endStream Whether this is a headers-only request.
   * @param idempotent indicates that the request is idempotent. When idempotent is set to true
   *   Envoy Mobile will retry on HTTP/3 post-handshake failures. By default, it is set to false.
   * @return This stream, for chaining syntax.
   */
  open fun sendHeaders(
    headers: RequestHeaders,
    endStream: Boolean,
    idempotent: Boolean = false
  ): Stream {
    underlyingStream.sendHeaders(headers.caseSensitiveHeaders(), endStream, idempotent)
    return this
  }

  /**
   * Read data from the response stream. Returns immediately.
   *
   * @param byteCount Maximum number of bytes that may be be passed by the next data callback.
   * @return This stream, for chaining syntax.
   */
  open fun readData(byteCount: Long): Stream {
    underlyingStream.readData(byteCount)
    return this
  }

  /**
   * For sending data to an associated stream. By default, the length sent is the
   * **[ByteBuffer.capacity]**. However, the length will rather be **[ByteBuffer.position]** if the
   * Stream was configured to do so - see **[StreamPrototype.useByteBufferPosition]**.
   *
   * Note: the provided ByteBuffer won't be mutated in any case. On the other hand, until the stream
   * is closed, any further mutations may lead to an unpredictable outcome.
   *
   * @param data Data to send over the stream.
   * @return This stream, for chaining syntax.
   */
  open fun sendData(data: ByteBuffer): Stream {
    var length = if (useByteBufferPosition) data.position() else data.capacity()
    underlyingStream.sendData(data, length, false)
    return this
  }

  /**
   * Close the stream with trailers.
   *
   * @param trailers Trailers with which to close the stream.
   */
  open fun close(trailers: RequestTrailers) {
    underlyingStream.sendTrailers(trailers.caseSensitiveHeaders())
  }

  /**
   * Close the stream with a data frame. By default, the length sent is the
   * **[ByteBuffer.capacity]**. However, the length will rather be **[ByteBuffer.position]** if the
   * Stream was configured to do so - see **[StreamPrototype.useByteBufferPosition]**.
   *
   * Note: the provided ByteBuffer won't be mutated in any case. On the other hand, until the stream
   * is closed, any further mutations may lead to an unpredictable outcome.
   *
   * @param data Data with which to close the stream.
   */
  open fun close(data: ByteBuffer) {
    var length = if (useByteBufferPosition) data.position() else data.capacity()
    underlyingStream.sendData(data, length, true)
  }

  /** Cancel the stream. */
  open fun cancel() {
    underlyingStream.cancel()
  }
}
